[Ang Kai Chao] iP

kaichaoang opened the pull request.

Commit:
Add Gradle support

kaichaoang committed:
Add the skeletal version of Duke

Let's start by greeting the user. Duke will echo commands entered by the user, and exits
when the user types the keyword "bye".

To enhance user experience, reserved keywords are added which the user can type
"help" to view all of those, and there are also visual lining breaks to provide a
clearer view of events.

kaichaoang committed:
Add ability to store user inputs

Let's add an ability to store past user inputs so that a user is able to
retrieve the contents of those should they be requested for.

kaichaoang committed:
Fix input bug

Scanner input type: white spaces break scanner inputs

Scanner input is fixed to accept one entry divided by whitespace

We want the scanner input to be able to detect more than just a single
word input to be more useful

Let's,
- change the scanner input type from .next() to .nextLine()

kaichaoang committed:
Add Task.java

Let's add a Task.java to adhere to java OOP principles. We should use the Task
to represent the user tasks instead of using the primitive String type.

kaichaoang committed:
Add the ability to check a task as done

Let's add the ability to check a task as done so that users can have a
better visual understanding of which tasks they have left.

Prompt the user if they are to provide an input that is not in line with
the required format.

kaichaoang committed:
Change if-else condition format

There were some corner cases for the "done" keyword that was found. For
the ease of splitting certain cases up, let's modify how conditions were being
checked, from directly using the user input to splitting it up via a white space.

kaichaoang committed:
Add Deadline.java

Let's add a Deadline.java to adhere to java OOP inheritance principles. We should use
the specific Deadline for deadline type tasks.

kaichaoang committed:
Add ToDo.java

Let's add a ToDo.java to adhere to java OOP inheritance principles. We should use
the specific ToDo for to-do type tasks.

kaichaoang committed:
Add Event.java

Let's add a Event.java to adhere to java OOP inheritance principles. We should use
the specific Event for event type tasks.

kaichaoang committed:
Add support for tracking 3 types of tasks

Let's add 3 new conditions to handle 3 special input types of tasks.
This will allow the user to see which types of specific tasks there is
remaining in the list and eases task handling for them.

We should also let user know the number of pending tasks there is
remaining.

Let's,
- add a line of output that allow users to know the number of tasks
  they have remaining after they added a new task.

kaichaoang committed:
Add support for tracking 3 types of tasks

Let's add 3 new conditions to handle 3 special input types of tasks.
This will allow the user to see which types of specific tasks there is
remaining in the list and eases task handling for them.

We should also let user know the number of pending tasks there is
remaining.

Let's,
- add a line of output that allow users to know the number of tasks
  they have remaining after they added a new task.
- add the 3 unqiue input types to our help keyword

kaichaoang committed:
Merge branch 'master' of https://github.com/kc-98/ip into master

kaichaoang committed:
Update bin folder files to compile

We've included new java files, so we have to compile all of them
instead of just the original Duke.java

kaichaoang committed:
Update Task.java status icon return

getStatusIcon(): return UTF-8 encoding \u2713 or \u2718

EXPECTED.TXT and ACTUAL.TXT are returning "?" instead of what
the UTF-8 encoding are supposed to return.

We want to manually type our UTF-8 encoding symbols to prevent
confusion.

Let's,
- change UTF-8 encoding \u2713 or \u2718 to "✓"  or  "✘" respectively

kaichaoang committed:
Fix Scanner NoSuchElementException

Let's add another terminating condition, scanner.hasNext().

This is to tell the scanner to stop expecting a new line if .hasNext()
returns false so that we do not encounter this error.

kaichaoang committed:
Add input testcases

Added some testcases in input.txt for testing purposes

kaichaoang committed:
Update expected results

Let's update our expected results after adding in the 3 special tasks for
testing purposes

kaichaoang committed:
Add InvalidKeyWordException.java

Let's add a Duke specific exception that is thrown to the user if they
input unknown keywords.

kaichaoang committed:
Add EmptyTaskException.java

Let's add a Duke specific exception that is thrown to the user if they
input a valid keyword followed by missing description.

kaichaoang committed:
Add DukeException.java

Let's add a Duke parent exception class so that our program can handle
Duke specific exceptions.

kaichaoang committed:
Update Duke input types and exceptions

Duke accepts any form of input

Duke might be buggy when faced with unsupported input formats.

Let's narrow Duke's input validity based on the given keywords to the
user via the help keyword and inform users should there be any invalid
inputs via exceptions

This can help the user plan their list in a more systematic structure
and minimise bug experienced.

Let's,
- restrict our keywords to add tasks only if supported keywords are used
 (as shown by inputting "help") .
- utilise Duke's customised exception classes wherever applicable to
  enhance UI

kaichaoang committed:
Add DeleteFailureException.java

Let's add a Duke specific exception that is thrown to the user if they
input an invalid deleting format or if there is an unexpected
deletion error.

kaichaoang committed:
Add deleting input type keyword

Let's add a delete keyword so that the user is able to delete
tasks off the existing list when they enter the correct delete
format as shown in under the help list.

We should also let user know the number of pending tasks there is
remaining after deletion

Let's,
- add a line of output that allow users to know the number of tasks
  they have remaining after they deleted an existing task.

kaichaoang committed:
Update java import statements

Let's explicitly state the java packages utilised to make the
class easier to comprehend and maintain.

kaichaoang committed:
Update java import statements

Let's explicitly state the java packages utilised to make the
class easier to comprehend and maintain.

kaichaoang committed:
Remove redunant class variable

Let's remove the class variable which is only used for the Task
of type Deadline and Event.

kaichaoang committed:
Add task saving and loading from hard-disk feature

Let's create a directory "./data" in our Duke folder, such that tasks
from previous sessions can be saved whenever the task list changes
and loaded when Duke boots up.

To support these features, let's:
-Include a parser method taskParser that reads a string and
 returns a specific task.

-Include a method loadData that creates a directory or the .txt file (if
 either are missing), or reads an existing file and returns a list of
 tasks used to store our existing tasks

-Include a method updateTaskFile that updates the list used to
 facilitate changes when the program is running

kaichaoang committed:
Modify class variable to LocalDate and LocalTime

Let's change the class String variable to LocalDate and LocalTime
variable so that Duke can understand the input as date and time
instead of just treating it as a String.

kaichaoang committed:
Modify class variable to LocalDate and LocalTime

Let's change the class String variable to LocalDate and LocalTime
variable so that Duke can understand the input as date and time
instead of just treating it as a String.

kaichaoang committed:
Update Duke to understand date and time

Let's update Duke to accepting stricter input format as shown
via the help keyword. This will help Duke to understand how
to read date and time inputs.

Lets,
- update taskParser and updateTaskFile methods to
  accomodate to using LocalDate and LocalTime
  instead of just plain String,
- add isValidDate and isValidTime to parse and determine
  if the user date and time input are valid to prevent any
  exceptions from being thrown by LocalDate and LocalTime
  classes.

kaichaoang committed:
Revert "Modify class variable to LocalDate and LocalTime"

This reverts commit 5cf1e3c90a13a88feea5238d8d76e35c69b7ffc7.

kaichaoang committed:
Remove redunant class variable

Let's remove the class variable which is only used for the Task
of type Deadline and Event.

kaichaoang committed:
Merge branch 'branch-Level-7' into master

# Conflicts:
# src/main/java/Duke.java

kaichaoang committed:
Modify class variable to LocalDate and LocalTime

Let's change the class String variable to LocalDate and LocalTime
variable so that Duke can understand the input as date and time
instead of just treating it as a String.

kaichaoang committed:
Change final variable naming style in Duke

Let's follow the Java convention and change the final
variale namings to all caps

kaichaoang committed:
Change final variable naming style in Duke

Let's follow the Java convention and capitalise all final
variable namings.

kaichaoang committed:
Merge branch 'branch-Level-8' into master

kaichaoang committed:
Update missing methods from branch-Level-8

There were some methods there were missing in the
master's Duke class after branch-Level-8 was merged
into.

kaichaoang committed:
Extract classes from Duke to make it more OOP

Let's extract classes from Duke to make our code to follow
the Java OOP principles, mainly the Single Responsibility
Principle.

Let's introduce the:
- Command class that has commands for each input type,
- Parser class that ensures input is readable by Duke,
- Storage class that handles any form of storage updates,
- UI class that handles any form of user interface situations.

kaichaoang committed:
Fix constructor lazy evaluation

The lazy evaluation in localDate and localTime variable
is wrong. "" should be checked instead of null so that it can be
parsed by Java LocalDate and LocalTime correctly

kaichaoang committed:
Add testcases in input.txt

Updated the testcases and expected outcomes. A
command to delete the tasks.txt file in Runtest.bat is
added so that testing is always done with a clean file.

kaichaoang committed:
Fix white space issue

The white space after the description segment of input
should be removed, and instead spacing should exist
after the descripition only there exist date and time.

kaichaoang committed:
Add JUnit test classes

Let's introduce some JUnit tests to check if the desired
outcomes are met. Let's target a few classes for now.

kaichaoang committed:
Remove redundant parameters and exception

Let's remove the parameters and exceptions that are
not needed in our methods.

kaichaoang committed:
Add InvalidFormatException.java

Let's add a Duke specific exception that is thrown to the user if they
input an invalid format that Duke does not understand.

kaichaoang committed:
Add Java Documentation

Add JavaDoc for non-simple methods so that Java is able
generate API documentation based on our JavaDocs.

kaichaoang committed:
Update hardwrap to 100 characters

Lets standardise to a coding style with hardwrap of 100
characters.

kaichaoang committed:
Add find feature to Duke

Let's add a find feature that allows user to be able
to find their tasks based on the task description only.

kaichaoang committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/Command.java
#	src/main/java/Parser.java
#	src/main/java/Storage.java
#	src/main/java/TaskList.java
#	src/main/java/Ui.java

kaichaoang committed:
Merge branch 'branch-Level-9' into master

# Conflicts:
#	src/main/java/Command.java

kaichaoang committed:
Fix missing add task feature

There was an accidental deletion of the method
addTaskCommand() in Command class, resulting in
Duke failing to add tasks.

Let's
- restore the addTaskCommand() under the execute() method
  in Command class

kaichaoang committed:
Merge remote-tracking branch 'origin/add-gradle-support' into master

kaichaoang committed:
Update application mainClassName to Duke

Let's change the mainClassName under application in
the build.gradle file to match the location of main method
which is in Duke.

kaichaoang committed:
Adhere to latest Java checkstyle

Let's correct our checkstyle errors to suit the latest
Java checkstyle.xml version 8.23.

kaichaoang committed:
Merge branch 'A-CheckStyle' into master

iamjamestan commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,81 @@
+        if (time.equals("")) {
+            return true;
+        }
+        String[] splitTime = time.split(":");
```
Perhaps it is better to rename "splitTime" to "periods"? Since according to the java coding standards, collection variables should be in plural form.

kaichaoang replied:
Thanks!

iamjamestan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,46 @@
+    public void run() {
+        command.welcomeCommand();
+        Scanner scanner = new Scanner(System.in);
+        boolean terminated = false;
```
Perhaps it may be more intuitive to rename the variable "terminated" to "isTerminated"?

kaichaoang replied:
Thanks, missed that one out!

iamjamestan reviewed (commented):
Overall consistent use of coding style. I like how you encapsulated related data into many different classes. It made the code easy to read and understand. LGTM, just a few nits with regards to coding standards to fix.

kaichaoang committed:
Incorporate GUI into Duke

Let's use Java FXML to add a GUI to Duke.

kaichaoang committed:
Fix minor code style issues

Changed some variable names that adhere to Java
coding standards.

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+        //determining user input type via the first word
+        String[] splitInput = splitUserInput(userInput);
+        String keyWord = getKeyWord(userInput);
+        if (keyWord.equals("welcome")) {
+            return welcomeCommand();
+        } else if (keyWord.equals("bye")) {
+            this.isTerminated = true;
+            return byeCommand();
+        } else if (keyWord.equals("help")) {
+            return helpCommand();
+        } else if (keyWord.equals("list")) {
+            return listCommand(tasks);
+        } else if (keyWord.equals("done")) {
+            return doneCommand(splitInput, tasks, storage);
+        } else if (keyWord.equals("delete")) {
+            return deleteCommand(tasks, splitInput, ui, storage);
+        } else if (keyWord.equals("todo") || keyWord.equals("deadline")
+                || keyWord.equals("event")) {
+            return addTaskCommand(splitInput, keyWord, tasks, storage);
+        } else if (keyWord.equals("find")) {
+            return findCommand(tasks, splitInput);
+        } else {
+            return ui.showErrorMessage(new DukeException("Unknown execution error.").getMessage());
+        }
```
Perhaps you can use a switch statement here instead of an if-else statement to improve code readability?

kaichaoang replied:
I can try that too, thanks!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+        this.isTerminated = false;
+    }
+
+    public boolean isTerminated() {
```
I like the naming of the boolean method. However, should this public method have a header comment? See [here](https://se-education.org/guides/conventions/java/index.html#comments). I noticed the same issue in several other places too.

kaichaoang replied:
Missed out on that standard, thanks!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+        return userInput.split(" ");
+    }
+
+    public String welcomeCommand() {
```
Should this method name be in the form of a verb instead? E.g. `printWelcomeCommand()`. I noticed the same issue in several other places too.

kaichaoang replied:
Okay will note, thanks!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+            EmptyTaskException {
+        Parser parser = new Parser();
+        if (array.length <= 1) {
+            if (array[0].equals("event")) {
+                throw new EmptyTaskException("☹ OOPS!!! The description of a event cannot "
+                        + "be empty.");
+            } else if (array[0].equals("deadline")) {
+                throw new EmptyTaskException("☹ OOPS!!! The description of a deadline cannot "
+                        + "be empty.");
+            } else if (array[0].equals("todo")) {
+                throw new EmptyTaskException("☹ OOPS!!! The description of a todo cannot "
+                        + "be empty.");
+            } else {
+                throw new InvalidFormatException("Invalid format error!");
+            }
```
Should you use a switch-case statement here instead?

kaichaoang replied:
Should be the same issue as a previous comment, will try it out thanks!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+            int pivotIndex = 0;
+            for (int i = 1; i < array.length; i++) {
+                if (array[i].equals("/at")) {
+                    //sets the breaking point of input
```
Should there be a spacing after "//" for comments? I noticed the same issue elsewhere too.

kaichaoang replied:
Missed that out thanks!!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+     * @throws InvalidFormatException If the input format is wrong.
+     * @throws EmptyTaskException If input task is not tagged with description.
+     */
+    public static String[] processUserTaskInput(String[] array) throws InvalidFormatException,
```
This method seems to be particularly long and quite convoluted. Perhaps you can consider the following methods to make it more readable:
1. Break the code into logic blocks and abstract them into separate methods.
2. Write more comments to explain what each logic block handles. 
3. Use more switch-case statements or else-if statements, instead of many nested if and else statements.

kaichaoang replied:
I noted the issue before and will resolve if time permits, thanks!

jeannetoh99 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,308 @@
+     *
+     * @param array Contains strings for task type, description, date/time if applicable.
+     * @throws InvalidFormatException If the input format is wrong.
+     * @throws EmptyTaskException If input task is not tagged with description.
```
Should there be a @return param here?

kaichaoang replied:
Oops missed out on that, thanks!

jeannetoh99 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public boolean hasDate() {
+        return this.localDate != null;
```
Should you remove the this keyword? The [java coding standard documentation](https://se-education.org/guides/conventions/java/index.html#variables) mentions to use this keyword only when a field is shadowed by a method or constructor parameter. I've noticed this issue in several other places as well.

kaichaoang replied:
Missed out on that standard, thanks!

jeannetoh99 reviewed (commented):
Generally great job! :) Maybe you can make use of switch-case statements / logic abstraction more to avoid too many nested if-else statements. Other than that, just a few nits here and there.

XIA-LIYI commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate localDate;
```
Hi, is there any special meaning you choose protected here.

kaichaoang replied:
I think I can change to private here, thanks!

XIA-LIYI commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,81 @@
+        if (hoursAndMinutes.length == 2) {
+            String hour = hoursAndMinutes[0];
+            String minutes = hoursAndMinutes[1];
+            if (hour.length() == 2 && minutes.length() == 2 && Integer.parseInt(hour) <= 24
```
Hi, I am not sure whether it is unnecessary to check whether the length is equal to 2.

kaichaoang replied:
Its to check if hour is within 0-24 range, but there's probably a better way for that, thanks!

XIA-LIYI commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+        this.taskList = tasks;
+    }
+
+    public int noOfTasks() {
```
Hi, I think it is better to name this method as getNoOfTasks.

kaichaoang replied:
Okay will do, thanks!

XIA-LIYI reviewed (commented):
Overall, the code is very nice and readable. Just some minor suggestions are given to you.

kaichaoang committed:
Fix more code style issues

Changed some method names that adhere to Java
coding standards.
Also, added getter methods after setting Event and
Deadline attributes to private.

kaichaoang committed:
Update shadow plugin version

kaichaoang committed:
Add assertions wherever possible

Let's add assertions at points of codes where it is not
supposed to be reached.

kaichaoang committed:
Improve code quality in code

Let's:
- Abstract long methods into shorter ones to improve
  readability,
- Use switch statements where applicable instead of if-else to
  for better readability,
- Add javadocs for all public methods.

kaichaoang committed:
Merge pull request #1 from kc-98/branch-A-CodeQuality

Improve code quality in code

kaichaoang committed:
Merge branch 'branch-A-Assertions' into master

# Conflicts:
#	src/main/java/Command.java

kaichaoang committed:
Add alphabetical task sorting feature

Let's add a sorting feature for Duke whereby tasks are
being sorted based on alphabetical order of their description.

kaichaoang committed:
Add the usage of enums for keywords

Let's use enums to improve our code quality.

kaichaoang committed:
Merge branch 'branch-A-CodeQuality' into master

# Conflicts:
#	src/main/java/Parser.java

kaichaoang committed:
Fix string to enum conversion bug

kaichaoang committed:
Use java streams for methods where applicable

kaichaoang committed:
Improve GUI of Duke

kaichaoang committed:
Set theme jekyll-theme-hacker

kaichaoang committed:
Set theme jekyll-theme-slate

kaichaoang committed:
Update README.md for Duke

Let's update the readme for duke and inlcude images
to guide users.

kaichaoang committed:
Merge branch 'master' of https://github.com/kc-98/ip into master

kaichaoang committed:
Update image quality for user guide

kaichaoang committed:
Set theme jekyll-theme-midnight

kaichaoang committed:
Fix minor exception message bug
